### ISSUE TITLE
New quote-props option for numbers

### ISSUE THREAD

**Issue by jabacchetta** [eslint/eslint#12135]
**What rule do you want to change?**
`quote-props`

**Does this change cause the rule to produce more or fewer warnings?**
Depends on configuration.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**

```js
const goodKeys = {
  // Good, valid (non-negative integer) array index
  1: '1',
};

// Problematic keys that are parsed and transformed before being typecasted into string
const badKeys = {
  // Interpreted as octal
  0o777: '511',

  // Interpreted as floating-point
  2.000: '2',

  // Interpreted as decimal
  000888: '888',
   
  // Interpreted as hexadecimal
  0x00111: '273',

  // Interpreted as binary
  0b0011: '3',
};
```

**What does the rule currently do for this code?**
The current `"numbers": true` option requires all numbers to have quotes.

**What will the rule do after it's changed?**
A new option that only requires numbers to have quotes if they are not valid (non-negative integer) array indexes.

**Are you willing to submit a pull request to implement this change?**
Not in the near future, but would be willing to eventually.
